programs/pluto/ikev2_parent.c

```diff
@@ -4215,7 +4228,82 @@ static bool ikev2_rekey_child_resp(struct ike_sa *ike, struct child_sa *child,
 	return true;
 }
 
-static bool ikev2_rekey_child_copy_ts(struct child_sa *child)
+/**
+ * ikev2_rekey_child_copy_sec_ctx: Copy the security label from the state of
+ * the predecessor child/IPsec SA to the state of the new child/IPsec SA
+ * replacing said predecessor during rekeying.
+ *
+ * @param[in,out]	new_state	State of the new child/IPsec SA.
+ * @param[in]		predecessor	State of the predecessor child/IPsec SA.
+ *
+ * @return	True if copying the security label from `predecessor` to
+ * 		`new_state` was successful.
+ */
+static bool ikev2_rekey_child_copy_sec_ctx(struct state *const new_state,
+					   struct state const* const predecessor) {
+	libreswan_log("%s: Prior to updating new SA's security label: predecessor security label = %s, new child/IPsec SA security label = %s",
+		     __func__,
+		     (predecessor->sec_ctx ? predecessor->sec_ctx->sec_ctx_value: "<NULL>"),
+		     (new_state->sec_ctx ? new_state->sec_ctx->sec_ctx_value : "<NULL>"));
+
+	if (predecessor->sec_ctx != NULL) {
+		/* Predecessor has a security label. */
+		if (new_state->sec_ctx != NULL) {
+			/*
+			 * New child/IPsec SA also has a security label already.
+			 * Check if it matches that of the predecessor.
+			 */
+			if (predecessor->sec_ctx == new_state->sec_ctx) {
+				return true;
+			}
+			if (streq(predecessor->sec_ctx->sec_ctx_value,
+				  new_state->sec_ctx->sec_ctx_value)) {
+				return true;
+			}
+
+			/* Security label mismatch between old and new child/IPsec SAs.*/
+			loglog(RC_LOG_SERIOUS, "ERROR: Security label mismatch between old SA (%s) and new SA (%s) during child rekeying", (predecessor->sec_ctx ? predecessor->sec_ctx->sec_ctx_value: "<NULL>"), (new_state->sec_ctx ? new_state->sec_ctx->sec_ctx_value : "<NULL>"));
+			return false;
+		} else {
+			/*
+			 * New child/IPsec SA doesn't have a security label.
+			 * Copy from the predecessor.
+			 */
+			new_state->sec_ctx = clone_thing(*predecessor->sec_ctx,
+							 "struct xfrm_user_sec_ctx_ike");
+			return true;
+		}
+	} else {
+		/* Predecessor does NOT have a security label. */
+		if (new_state->sec_ctx != NULL) {
+			/*
+			 * New child/IPsec SA has a security label while the
+			 * predecessor doesn't. This should not happen -
+			 * possible bug.
+			 */
+			loglog(RC_LOG_SERIOUS, "ERROR: Security label mismatch between old SA (<NULL>) and new SA (%s) during child rekeying", new_state->sec_ctx->sec_ctx_value);
```

Comment:
would be nice to split this log line into to lines.

Reply:
Fixed.